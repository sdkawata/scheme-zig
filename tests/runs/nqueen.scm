(define range 
    (letrec ((range_tail (lambda (start end acc)
        (if (= start end) acc (range_tail start (- end 1) (cons (- end 1) acc))))))
        (lambda (start end) (range_tail start end (quote ())))))
(define find_first (lambda (f l)
    (if 
        (null? l)
        #f
        (let ((r (f (car l))))
            (if r r (find_first f (cdr l)))))))
(define any
    (lambda (f l)
        (if (null? l)
            #f
            (if (f (car l)) #t (any f (cdr l))))))
(define nqueen
    (lambda (n)
        (letrec ((rec (lambda (cy acc)
            (if
                (= cy n)
                acc
                (find_first 
                    (lambda (cx)
                        (if 
                            (any
                                (lambda (pos) (let ((px (car pos)) (py (car (cdr pos))))
                                    (or (= px cx) (= (- py cy) (- px cx)) (= (- py cy) (- cx px)))))
                                acc)
                            #f
                            (rec (+ cy 1) (cons (cons cx (cons cy (quote ()))) acc))))
                    (range 0 n))))))
            (rec 0 (quote ())))))
(nqueen 8)