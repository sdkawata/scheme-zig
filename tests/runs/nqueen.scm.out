((3 7) (1 6) (6 5) (2 4) (5 3) (7 2) (4 1) (0 0))